pkg/gadgets/network-graph/gadget.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2019-2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package networkgraph
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+	containerutils "github.com/kinvolk/inspektor-gadget/pkg/container-utils"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets"
+	nettracer "github.com/kinvolk/inspektor-gadget/pkg/gadgets/network-graph/tracer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/network-graph/types"
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
+
+type pubSubKey string
+
+type Trace struct {
+	resolver gadgets.Resolver
+	client   client.Client
+
+	started bool
+	done    chan bool
+
+	tracer   *nettracer.Tracer
+	enricher *Enricher
+	wg       sync.WaitGroup
+
+	netnsHost uint64
+
+	pubSubKey pubSubKey
+}
+
+type TraceFactory struct {
+	gadgets.BaseFactory
+
+	netnsHost uint64
+}
+
+func NewFactory() gadgets.TraceFactory {
+	netnsHost, _ := containerutils.GetNetNs(os.Getpid())
```

Comment:
Isn't it possible to do this inside the [`n` function](https://github.com/kinvolk/inspektor-gadget/blob/55468a486cad05d6d4528450dd846e5373862016/pkg/gadgets/network-graph/gadget.go#L88-L94) and avoid carrying it inside `TraceFactory`?

Reply:
Possible, but that would re-execute the lookup in `/proc/$GADGET_PID/ns/net` for each new trace. With the TraceFactory, it is done only one time.